src/backend/aspen/workflows/nextstrain_run/create_phyloruns.py

```diff
@@ -129,6 +135,121 @@ def launch_one(
+            db, group, pathogen, repo, location, filter_start_date, filter_end_date
+        )
+        if args:
+            print("match found")
+            return args
+        print("no matches, trying again")
+
+        # Try going back further in time.
+        if start_date_attempts:
+            filter_start_date = dateparser.parse(start_date_attempts.pop(0)).date()  # type: ignore
+            # We overwrote the start date to look further back in time, let's try again!
+            continue
+
+        # widen our geographical search area
+        if (
+            "division" in current_location_level
+        ):  # We don't build scheduled trees for locations bigger than a country.
+            current_location_level = current_location_level[:-1]
+            new_location_query = sa.select(Location)
+            for col in location_hierarchy:
+                col_match = None
+                if col in current_location_level:
+                    col_match = getattr(location, col)
+                new_location_query = new_location_query.where(
+                    getattr(Location, col) == col_match
+                )
+            location = db.execute(new_location_query).scalars().one()
+            # We overwrote the location var with a wider area. Try again!
+            continue
+
+        # If we've made it this far, we have no more options, skip the build..
+        return None
+
+
+def get_template_args_for_focal_group(
+    db,
+    group: Group,
+    pathogen: Pathogen,
+    repo: PublicRepository,
+    location: Location,
+    filter_start_date: datetime.date,
+    filter_end_date: datetime.date,
+):
+    template_args = {
+        "location_id": location.id,
+        "filter_start_date": filter_start_date.strftime("%Y-%m-%d"),
+        "filter_end_date": filter_end_date.strftime("%Y-%m-%d"),
+    }
+
+    # If this group has uploaded samples within our start/end dates, we're all set.
+    group_samples = db.execute(
+        sa.select([func.count()])
+        .select_from(Sample)
+        .where(
+            Sample.pathogen == pathogen,
+            Sample.submitting_group == group,
+            Sample.collection_date >= filter_start_date,
+            Sample.collection_date <= filter_end_date,
+        )
+    ).scalar()
+    if group_samples:
+        return template_args
+
+    # Even if the group hasn't uploaded samples, we can proceed if anyone's uploaded
+    # data to gisaid for their default tree location.
+    upstream_samples_query = (
+        sa.select([func.count()])
+        .select_from(PublicRepositoryMetadata)
+        .where(
+            PublicRepositoryMetadata.region == location.region,
+            PublicRepositoryMetadata.country == location.country,
+            PublicRepositoryMetadata.pathogen == pathogen,
+            PublicRepositoryMetadata.public_repository == repo,
+            PublicRepositoryMetadata.date >= filter_start_date,
+            PublicRepositoryMetadata.date <= filter_end_date,
```

Comment:
just out of curiosity, is there a things such as return TRUE if the first match is found, rather than finding all matches? or this amount of time is not worth saving?

Reply:
You're actually right -- here we're fetching a count() of rows, and we could instead try to get the first match which (depending on a bunch of internal PostgreSQL voodoo) is probably a bit faster. I've updated the pr!